src/util/ActionProgramUtils.ts

```diff
@@ -238,109 +238,153 @@ export const makeAaveCloseAction = async ({
+  const { currencyCode: collateralCurrencyCode, denominations: collateralDenoms } =
+    collateralTokenId == null ? currencyInfo : allTokens[collateralTokenId] ?? {}
 
-  // Withdraw actions
+  // Enable withdrawal token in case somehow disabled after opening LoanAccunt
   const withdrawalToken = getToken(wallet, collateralTokenId)
   if (withdrawalToken != null) {
-    // Make sure the collateralized token is enabled just because
     await enableToken(withdrawalToken.currencyCode, wallet)
-
-    seqAction.actions.push({
-      type: 'loan-withdraw',
-      borrowPluginId,
-      nativeAmount: MAX_AMOUNT.toString(),
-      tokenId: collateralTokenId,
-      walletId: wallet.id
-    })
   }
 
-  if (seqAction.actions.length > 0) {
-    return seqAction
+  // Handle any debts, if they exist
+  const debt = debts[0]
+  if (debt.tokenId != null) {
+    // Enable debt token in case somehow disabled after opening LoanAccount
+    const debtTokenId = debt.tokenId
+    const { currencyCode: debtCurrencyCode, denominations: debtDenoms } = allTokens[debtTokenId] ?? {}
+    await enableToken(debtCurrencyCode, wallet)
+
+    // #region Swap Validation
+
+    const collateralDenom = collateralDenoms[0]
+    const collateralMultiplier = collateralDenom?.multiplier ?? '1'
+    const collateralFiat = convertCurrencyFromExchangeRates(
+      exchangeRates,
+      collateralCurrencyCode,
+      isoFiatCurrencyCode,
+      convertNativeToExchange(collateralMultiplier)(collateral.nativeAmount)
+    )
+    const debtDenom = debtDenoms[0]
+    const debtMultiplier = debtDenom?.multiplier ?? '1'
+    const principalFiat = convertCurrencyFromExchangeRates(
+      exchangeRates,
+      debtCurrencyCode,
+      isoFiatCurrencyCode,
+      convertNativeToExchange(debtMultiplier)(debt.nativeAmount)
+    )
+    const debtBalanceNativeAmount = wallet.balances[debtCurrencyCode]
+    const debtBalanceFiat = convertCurrencyFromExchangeRates(
+      exchangeRates,
+      debtCurrencyCode,
+      isoFiatCurrencyCode,
+      convertNativeToExchange(debtMultiplier)(debtBalanceNativeAmount)
+    )
+
+    // First, calculate the remaining principal after a payment is made from the
+    // debt asset balance
+    const postRepayPrincipalFiat = sub(principalFiat, debtBalanceFiat)
+
+    // Find maximum amount of collateral that can be swapped while remaining
+    // above the liquidation threshold (74%), where:
+    // Pr: Principal remaining after initial payment above
+    // C: Collateral
+    // Cs: Collateral to swap (maxSwappableCollateral)
+    //
+    // Pr / (C - Cs) = 0.74
+    // Solved for Cs:
+    // Cs = (-1/.74) * Pr + C
+    const maxSwappableCollateralFiat = add(mul(div('-1', '0.74', DECIMAL_PRECISION), postRepayPrincipalFiat), collateralFiat)
+
+    // Check if swap and repay amount necessary to cover the remaining principal
+    // would cause liquidation
+    if (gt(postRepayPrincipalFiat, maxSwappableCollateralFiat)) {
```

Comment:
The amount shown here is not right, it's greater than what's actually needed. Not sure what the right maths are to get that value yet... Might be a calculus thing

Reply:
https://app.asana.com/0/1200382638405084/1203374650739508/f